qutip/visualization.py

```diff
@@ -868,66 +909,107 @@ def matrix_histogram(M, x_basis=None, y_basis=None, limits=None,
 
-        artist = ax.bar3d(xpos, ypos, zpos, dx, dy, bar_M, color=colors,
-                          edgecolors=options['bars_edgecolor'],
-                          linewidths=options['bars_lw'],
-                          shade=options['shade'])
-        artist_list.append([artist])
+        temp_xpos = xpos.reshape(M.shape)
+        temp_ypos = ypos.reshape(M.shape)
+        temp_zpos = zpos.reshape(M.shape)
+
+        # calculating z_order for each bar based on its position
+        z_order = (
+            np.multiply(
+                [
+                    temp_xpos, temp_ypos, temp_zpos], camera
+                    ).sum(0).flatten()
+        )
+
+        for i, uxpos in enumerate(xpos):
+            bars.append(
+                [
+                    uxpos,
+                    ypos[i],
+                    zpos[i],
+                    dx[i],
+                    dy[i],
+                    bar_M[i],
+                    colors[i],
+                    z_order[i]
+                ]
+            )
+
+        # Adding bars individually to the plot
+        for bar in bars:
+            artist = ax.bar3d(
+                bar[0],
+                bar[1],
+                bar[2],
+                bar[3],
+                bar[4],
+                bar[5],
+                color=bar[6],
+                edgecolors=options["bars_edgecolor"],
+                linewidths=options["bars_lw"],
+                shade=options["shade"],
+            )
+            # sorting the bars based on our calculated z_order
+            artist._sort_zpos = bar[7]
+            artist_list.append([artist])
+        # The sorting issue was fixed by making minor change to
+        # https://stackoverflow.com/questions/18602660/matplotlib-bar3d-clipping-problems
 
     if len(Ms) == 1:
         output = ax
     else:
-        output = animation.ArtistAnimation(fig, artist_list, interval=50,
-                                           blit=True, repeat_delay=1000)
+        output = animation.ArtistAnimation(
+            fig, artist_list, interval=50, blit=True, repeat_delay=1000
+        )
 
     # remove vertical lines on xz and yz plane
-    ax.yaxis._axinfo["grid"]['linewidth'] = 0
-    ax.xaxis._axinfo["grid"]['linewidth'] = 0
+    ax.yaxis._axinfo["grid"]["linewidth"] = 0
+    ax.xaxis._axinfo["grid"]["linewidth"] = 0
 
     # x axis
-    _update_xaxis(options['bars_spacing'], M, ax, x_basis)
+    _update_xaxis(options["bars_spacing"], M, ax, x_basis)
 
     # y axis
-    _update_yaxis(options['bars_spacing'], M, ax, y_basis)
+    _update_yaxis(options["bars_spacing"], M, ax, y_basis)
 
     # z axis
-    _update_zaxis(ax, z_min, z_max, options['zticks'])
+    _update_zaxis(ax, z_min, z_max, options["zticks"])
 
     # stick to xz and yz plane
-    _stick_to_planes(options['stick'],
-                     options['azim'], ax, M,
-                     options['bars_spacing'])
-    ax.view_init(azim=options['azim'], elev=options['elev'])
```

Comment:
Is this line removed on purpose?

Reply:
Yes , it resets the view once the bars were created and zpos was determined which becomes a problem in the ordering, this too was part of the issue hence removed